src/storage/src/memtable/btree.rs

```diff
@@ -196,25 +204,53 @@ impl BTreeIterator {
         let key_columns = rows_to_vectors(
             key_data_types,
             self.adapter.source_key_needed(),
+            &row_selection,
             keys.as_slice(),
         );
         let field_columns = rows_to_vectors(
             value_data_types,
             self.adapter.source_value_needed(),
+            &row_selection,
             values.as_slice(),
         );
 
+        let sequences = arrow::compute::take(&sequences.to_arrow_array(), &row_selection, None)
+            .context(error::SortArraysSnafu)?;
+        let op_type = arrow::compute::take(&op_types.to_arrow_array(), &row_selection, None)
+            .context(error::SortArraysSnafu)?;
+
         let batch = self.adapter.batch_from_parts(
             key_columns,
             field_columns,
-            Arc::new(sequences),
-            Arc::new(op_types),
+            Helper::try_into_vector(sequences)
+                .context(error::ConvertChunkSnafu { name: "sequence" })?,
+            Helper::try_into_vector(op_type)
+                .context(error::ConvertChunkSnafu { name: "op_type" })?,
         )?;
 
         Ok(Some(batch))
     }
 }
 
+fn filter_keys(keys: &[&InnerKey], range: Option<&TimestampRange>) -> UInt32Array {
```

Comment:
We have a `MapIterWrapper` to filter the entries by sequence. Maybe we can filter the entries by time range (if any) in it.

https://github.com/GreptimeTeam/greptimedb/blob/4ae0b5e1855ba37e0007e9f0c55fc5577b200c06/src/storage/src/memtable/btree.rs#L260-L268